gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/op/traversal/TraversalOpProcessor.java

```diff
@@ -383,9 +383,14 @@ protected void handleIterator(final Context context, final Iterator itty, final
                         break;
                     }
 
+                    // track whether there is anything left in the iterator because it needs to be accessed after
+                    // the transaction could be closed - in that case a call to hasNext() could open a new transaction
+                    // unintentionally
+                    final boolean moreInIterator = itty.hasNext();
```

Comment:
Instead of creating a new variable, can we reset the value of existing variable `hasMore` here?

Reply:
`hasMore` is only set to `false` in line 401 and that is in the `else` block of the if condition: `if (moreInIterator)`. So if we also use `hasMore` here instead, then `hasMore` cannot become `false` any more.
This additional `itty.hasNext()` is needed here to check whether more elements are available after the last `next()`. This is also explained in the comment above starting at line 337:

```
            // there is a need to check hasNext() on the iterator because if the channel is not writeable the
            // previous pass through the while loop will have next()'d the iterator and if it is "done" then a
            // NoSuchElementException will raise its head. also need a check to ensure that this iteration doesn't
            // require a forced flush which can be forced by sub-classes.
```

This is also in line with the implementation in `AbstractOpProcessor` where it's implemented exactly like this. We really just copied the logic from there.

Reply:
Thanks for the detailed explanation! Our understanding of your earlier suggestion was wrong based on that. Our assumption was that you wanted to get rid of the additional call `itty.hasNext()` completely.

But then it definitely sounds good to reuse `hasMore` hereinstead of adding `moreInIterator`. Pushed an updated commit that incorporates this change also for the `AbstractOpProcessor` which already used `moreInIterator` before.